### PR TITLE
#1404 Added Message History to Serial Monitor

### LINKED ISSUE #1404: Serial Monitor message history
### Describe the request

In Arduino IDE 1.x, you can 'up arrow' through the recently sent commands, to send the same one a second time, or even go further back in history.

### Describe the current behavior

none, this functionality is missing.

### Arduino IDE version

IDE 2.0.0-rc9.3

### Operating system

macOS

### Operating system version

Monterey

### Additional context

_No response_

### Issue checklist

- [X] I searched for previous requests in [the issue tracker](https://github.com/arduino/arduino-ide/issues?q=)
- [X] I verified the feature was still missing when using the latest [nightly build](https://github.com/arduino/arduino-ide#nightly-builds)
- [X] My request contains all necessary details

### PR DESCRIPTION
**Pull request by nmzaheer**
### Motivation
Feature added for message history in Serial Monitor

### Change description
Utilizes the RingList implementation of @dwightfowler-cd  [here](https://github.com/dwightfowler-cd/arduino-ide/blob/Issue-1404/arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx) with some corrections for traversal of the list

### Other information
- Another PR for this is available at PR #1467
- This closes #1404 
### Reviewer checklist

* [x] PR addresses a single concern.
* [ ] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-ide/pulls) before creating one)
* [x] PR title and description are properly filled.
* [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

**Comment by nmzaheer** [comment]
There was an error in the next method wherein the last value was being shown twice. It has been corrected with the latest commit. I have also downloaded the artifact and tested against it.

**Comment by dwightfowler-cd** [comment]
Akos,
I believe there is still a bug in the way it behaves. I have not had a chance to get back to debugging the code.

Is the team getting close to a release?


Thanks,


Dwight

**Comment by kittaakos** [comment]
> Is the team getting close to a release?

I will ensure this feature is part of the next patch release of IDE2. We plan to roll out the next version this month. Please bear with us.

@dwightfowler-cd, if you do not plan to do follow-ups on #1467, please close it. Thanks for keeping the repo clean ❤️

**Comment by nmzaheer** [comment]
@kittaakos Did you have a chance to test the behaviour? Do you have any feedback for me?

**Comment by kittaakos** [comment]
It's working great. Thank you for your contribution. I believe the logic can be simplified. Let's remove the unnecessary state from the `HistoryList`. See my proposal [here](https://github.com/nmzaheer/arduino-ide/compare/nmzaheer-issue-1404...kittaakos:arduino-ide:%231404-review). What do you think?

Please give it a try. If you like it, you can [`cherry-pick`](https://git-scm.com/docs/git-cherry-pick) the changes, or you can add my fork to your remotes and pull in the changes.

**Comment by nmzaheer** [comment]
Thank you for testing. Your implementation looks much neater and simpler than what I had submitted. 

I am testing your implementation of HistoryList with [this](https://gist.github.com/nmzaheer/99785c5ec7ce5be132d32adee94d858c) code. However, it skips the first entry when `next` method is run. I will have to correct the code accordingly.

I'll get back to you with the changes.

File: arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx [hunk 1]
```diff
@@ -110,8 +158,18 @@ export class SerialMonitorSendInput extends React.Component<
...
+        this.setState({ text: this.state.history.previous() });
+      } else if (key === Key.ARROW_DOWN) {
+        this.setState({ text: this.state.history.next() });
+      } else if (key === Key.ESCAPE) {
+        this.setState({ text: '' });
       }
     }
   }
-}
+}
```
Comment (kittaakos) [inline arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx:175]: _No newline at end of file_

Please fix it.

**Review by kittaakos — COMMENTED** [review]
The code is looking good and the monitor history works as in IDE 1.x 👍

File: arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx (hunk 1 above)
Comment (nmzaheer) [reply arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx]: Newline added.

**Review by kittaakos — APPROVED** [review]
Thank you! It's looking good.

**Pull request merged**